File: adventofcode/2021/2/kotlin/2.kt
fun slv(actions: List<Pair<String, Int>>): Int {
    val (horizontalPosition, depth) = actions.partition {
        it.first == "forward"
    }.let {
        (forwards, upDown) -> Pair(
            forwards.map { (_, value) -> value }.sum(),
            upDown.map { (direction, value) -> if (direction == "up") -value else value }.sum()
        )
    }
    return horizontalPosition * depth
}

fun slv2(actions: List<Pair<String, Int>>) : Int {
    var aim = 0;
    val depth = actions.map {
        (direction, value) -> when(direction) {
            "down" -> {aim += value; 0}
            "up" -> {aim -= value; 0}
            else -> aim*value
        }
    }.sum()
    val horizontalPosition = actions.filter { (direction, _) -> direction == "forward" }.map{it.second}.sum()
    return depth * horizontalPosition
}

fun main() {
    val actions = (0 until readLine()!!.toInt()).map { readLine()!!.split(" ").let { (direction, number) -> Pair(direction, number.toInt()) } }
    println("Problem 1: ${slv(actions)}")
    println("Problem 2: ${slv2(actions)}")
}

File: adventofcode/2021/2/c++/2.cpp
#include <bits/stdc++.h>
using namespace std;

int slv(vector<pair<string, int> > a, int n) {
    int d, h;
    d = h = 0;
    for(int i = 0; i<n; i++) {
        switch (a[i].first[0]) {
            case 'u': d -= a[i].second; break;
            case 'd': d += a[i].second; break;
            case 'f': h += a[i].second; break;
        }
    }
    return d*h;
}

int slv2(vector<pair<string, int> > a, int n) {
    int d, h, aim;
    d = h = aim = 0;
    for(int i = 0; i<n; i++) {
        switch (a[i].first[0]) {
            case 'u': aim -= a[i].second; break;
            case 'd': aim += a[i].second; break;
            case 'f': h += a[i].second; d += aim * a[i].second; break;
        }
    }
    return d*h;
}

int main() {
    int n;
    cin >> n;
    vector<pair<string, int> > a(n);
    for (int i = 0; i < n; i++)
        cin >> a[i].first >> a[i].second;
    cout << "Problem1: " << slv(a, n) << endl;
    cout << "Problem2: " << slv2(a, n) << endl;
}